runtime/caml/platform.h

```diff
@@ -452,6 +502,45 @@ CAMLextern CAMLthread_local int caml_lockdepth;
 #define DEBUG_UNLOCK(m)
 #endif
 
+#ifdef _WIN32
+
+Caml_inline void caml_plat_lock_blocking(caml_plat_mutex* m)
+{
+  DWORD self_tid = GetCurrentThreadId();
+  if (m->owner_tid != self_tid) {
+    AcquireSRWLockExclusive(&m->lock);
+    m->owner_tid = self_tid;
+    DEBUG_LOCK(m);
+  } else {
+    check_err("lock", EDEADLK);
+  }
+}
```

Comment:
It seems unfortunate that `ERRORCHECK`-related code should be essentially duplicated in three different places (here, `sync_mutex_lock`, and `caml_plat_lock_non_blocking_actual`, and others). Surely they can be extracted out to functions that directly mirror the pthreads behaviour? e.g.

```c
Caml_inline int caml_plat_lock_errorcheck(caml_plat_mutex* m)
{
  DWORD self_tid = GetCurrentThreadId();
  if (m->owner_tid != self_tid) {
    AcquireSRWLockExclusive(&m->lock);
    m->owner_tid = self_tid;
    return 0;
  } else {
    return EDEADLK;
  }
}
```

The code isn't entirely trivial: it took me a second to convince myself that the `m->owner_tid` check is _not_ a time-of-check to time-of-use error, for instance.

Reply:
I agree with you, but considering that the duplication between sync and plat modules existed before this PR, and although I'm tempted to try to merge the code, I'm reluctant to do it here or at all (and I would appreciate a greenlight from a core dev before diving in).

Reply:
As the duplication existed before this PR, let us leave this potential improvement for a separate PR as well, to keep this one as focused as possible.